File: Core kernel razonamiento repo para Yaiwes/GitHub-Copilot-SDK/GitHub-Copilot-SDK/nodejs/src/sessionFsProvider.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import type {
    SessionFsHandler,
    SessionFsError,
    SessionFsStatResult,
    SessionFsReaddirWithTypesEntry,
    SessionFsSqliteQueryResult as GeneratedSqliteQueryResult,
    SessionFsSqliteTransactionError as GeneratedSqliteTransactionError,
    SessionFsSqliteTransactionErrorClass,
    SessionFsSqliteQueryType,
} from "./generated/rpc.js";

export type { SessionFsSqliteQueryType, SessionFsSqliteTransactionErrorClass };

/**
 * File metadata returned by {@link SessionFsProvider.stat}.
 * Same shape as the generated {@link SessionFsStatResult} but without the
 * `error` field, since providers signal errors by throwing.
 */
export type SessionFsFileInfo = Omit<SessionFsStatResult, "error">;

/**
 * Result of a SQLite query execution via {@link SessionFsSqliteProvider.query}.
 * Same shape as the generated {@link GeneratedSqliteQueryResult} but without the
 * `error` field, since providers signal errors by throwing.
 */
export type SessionFsSqliteQueryResult = Omit<GeneratedSqliteQueryResult, "error">;

/**
 * One statement in an atomic SQLite transaction passed to
 * {@link SessionFsSqliteProvider.transaction}.
 */
export interface SessionFsSqliteStatement {
    /** How to execute: `"exec"` for DDL/multi-statement, `"query"` for SELECT, `"run"` for INSERT/UPDATE/DELETE. */
    queryType: SessionFsSqliteQueryType;

    /** SQL statement to execute. */
    query: string;

    /** Optional named bind parameters. */
    params?: Record<string, string | number | null>;
}

/**
 * Error thrown by {@link SessionFsSqliteProvider.transaction} to classify a
 * transaction failure for the runtime.
 *
 * Any other thrown value is reported as `"fatal"`. Throw this with
 * `"busyOrLocked"` when SQLite reported BUSY/LOCKED before commit and the
 * transaction was rolled back, so the runtime knows the call is safe to retry.
 */
export class SessionFsSqliteTransactionFailure extends Error {
    /** Failure classification reported to the runtime. */
    readonly errorClass: SessionFsSqliteTransactionErrorClass;

    constructor(message: string, errorClass: SessionFsSqliteTransactionErrorClass = "fatal") {
        super(message);
        this.name = "SessionFsSqliteTransactionFailure";
        this.errorClass = errorClass;
    }
}

/**
 * SQLite operations for the per-session database.
 * Implementers provide query execution and existence checking.
 */
export interface SessionFsSqliteProvider {
    /**
     * Execute a SQLite query against the per-session database.
     *
     * @param queryType - How to execute: `"exec"` for DDL/multi-statement, `"query"` for SELECT, `"run"` for INSERT/UPDATE/DELETE.
     * @param query - SQL query to execute.
     * @param params - Optional named bind parameters.
     */
    query(
        queryType: SessionFsSqliteQueryType,
        query: string,
        params?: Record<string, string | number | null>
    ): Promise<SessionFsSqliteQueryResult | undefined>;

    /**
     * Execute `statements` atomically against the per-session database.
     *
     * Apply busy handling to every statement and roll back the whole batch if
     * any statement fails. Throw {@link SessionFsSqliteTransactionFailure} to
     * classify the failure; any other thrown value is reported as `"fatal"`.
     *
     * @param statements - Statements to execute in order inside a single transaction.
     * @returns One result per statement, in the same order.
     */
    transaction?(statements: SessionFsSqliteStatement[]): Promise<SessionFsSqliteQueryResult[]>;

    /**
     * Check whether the per-session database already exists, without creating it.
     */
    exists(): Promise<boolean>;
}

/**
 * Interface for session filesystem providers. Implementers use idiomatic
 * TypeScript patterns: throw on error, return values directly. Use
 * {@link createSessionFsAdapter} to convert a provider into the
 * {@link SessionFsHandler} expected by the SDK.
 *
 * Errors with a `code` property of `"ENOENT"` are mapped to the ENOENT
 * error code; all others map to UNKNOWN.
 */
export interface SessionFsProvider {
    /** Reads the full content of a file. Throw if the file does not exist. */
    readFile(path: string): Promise<string>;

    /** Writes content to a file, creating parent directories if needed. */
    writeFile(path: string, content: string, mode?: number): Promise<void>;

    /** Appends content to a file, creating parent directories if needed. */
    appendFile(path: string, content: string, mode?: number): Promise<void>;

    /** Checks whether a path exists. */
    exists(path: string): Promise<boolean>;

    /** Gets metadata about a file or directory. Throw if it does not exist. */
    stat(path: string): Promise<SessionFsFileInfo>;

    /** Creates a directory. If recursive is true, creates parents as needed. */
    mkdir(path: string, recursive: boolean, mode?: number): Promise<void>;

    /** Lists entry names in a directory. Throw if it does not exist. */
    readdir(path: string): Promise<string[]>;

    /** Lists entries with type info. Throw if the directory does not exist. */
    readdirWithTypes(path: string): Promise<SessionFsReaddirWithTypesEntry[]>;

    /** Removes a file or directory. If force is true, do not throw on ENOENT. */
    rm(path: string, recursive: boolean, force: boolean): Promise<void>;

    /** Renames/moves a file or directory. */
    rename(src: string, dest: string): Promise<void>;

    /** Per-session SQLite database operations. Optional — omit if the provider does not support SQLite. */
    sqlite?: SessionFsSqliteProvider;
}

function normalizeSqliteParams(
    params?: Record<string, unknown>
): Record<string, string | number | null> | undefined {
    if (!params) {
        return undefined;
    }

    const normalized: Record<string, string | number | null> = {};
    for (const [key, value] of Object.entries(params)) {
        if (value !== undefined) {
            normalized[key] = value as string | number | null;
        }
    }
    return normalized;
}

/**
 * Wraps a {@link SessionFsProvider} into the {@link SessionFsHandler}
 * interface expected by the SDK, converting thrown errors into
 * {@link SessionFsError} results.
 */
export function createSessionFsAdapter(provider: SessionFsProvider): SessionFsHandler {
    return {
        readFile: async ({ path }) => {
            try {
                const content = await provider.readFile(path);
                return { content };
            } catch (err) {
                return { content: "", error: toSessionFsError(err) };
            }
        },
        writeFile: async ({ path, content, mode }) => {
            try {
                await provider.writeFile(path, content, mode);
                return undefined;
            } catch (err) {
                return toSessionFsError(err);
            }
        },
        appendFile: async ({ path, content, mode }) => {
            try {
                await provider.appendFile(path, content, mode);
                return undefined;
            } catch (err) {
                return toSessionFsError(err);
            }
        },
        exists: async ({ path }) => {
            try {
                return { exists: await provider.exists(path) };
            } catch {
                return { exists: false };
            }
        },
        stat: async ({ path }) => {
            try {
                return await provider.stat(path);
            } catch (err) {
                return {
                    isFile: false,
                    isDirectory: false,
                    size: 0,
                    mtime: new Date().toISOString(),
                    birthtime: new Date().toISOString(),
                    error: toSessionFsError(err),
                };
            }
        },
        mkdir: async ({ path, recursive, mode }) => {
            try {
                await provider.mkdir(path, recursive ?? false, mode);
                return undefined;
            } catch (err) {
                return toSessionFsError(err);
            }
        },
        readdir: async ({ path }) => {
            try {
                const entries = await provider.readdir(path);
                return { entries };
            } catch (err) {
                return { entries: [], error: toSessionFsError(err) };
            }
        },
        readdirWithTypes: async ({ path }) => {
            try {
                const entries = await provider.readdirWithTypes(path);
                return { entries };
            } catch (err) {
                return { entries: [], error: toSessionFsError(err) };
            }
        },
        rm: async ({ path, recursive, force }) => {
            try {
                await provider.rm(path, recursive ?? false, force ?? false);
                return undefined;
            } catch (err) {
                return toSessionFsError(err);
            }
        },
        rename: async ({ src, dest }) => {
            try {
                await provider.rename(src, dest);
                return undefined;
            } catch (err) {
                return toSessionFsError(err);
            }
        },
        // Unlike the FS methods above, SQLite methods let errors propagate to the JSON-RPC layer
        // rather than catching and mapping via toSessionFsError. The FS error mapping is specifically
        // for translating Node.js errno codes (e.g., ENOENT) into SessionFsError, which isn't
        // meaningful for SQL errors. Letting exceptions propagate preserves the original error
        // message in the JSON-RPC error response.
        sqliteQuery: async ({ queryType, query, params: bindParams }) => {
            if (!provider.sqlite) {
                throw new Error("SQLite is not supported by this provider");
            }
            const result = await provider.sqlite.query(
                queryType,
                query,
                normalizeSqliteParams(bindParams)
            );
            return result ?? { rows: [], columns: [], rowsAffected: 0 };
        },
        sqliteTransaction: async ({ statements }) => {
            if (!provider.sqlite?.transaction) {
                return {
                    results: [],
                    error: {
                        errorClass: "fatal",
                        message: "SQLite transactions are not supported by this provider",
                    },
                };
            }
            try {
                const results = await provider.sqlite.transaction(
                    statements.map((statement) => ({
                        queryType: statement.queryType,
                        query: statement.query,
                        params: normalizeSqliteParams(statement.params),
                    }))
                );
                return { results: results.map((result) => ({ ...result })) };
            } catch (err) {
                // Unlike sqliteQuery, transaction failures carry a classification the
                // runtime uses to decide whether a retry is safe, so they are reported
                // as a result-level error instead of a JSON-RPC error.
                return { results: [], error: toSqliteTransactionError(err) };
            }
        },
        sqliteExists: async () => {
            if (!provider.sqlite) {
                throw new Error("SQLite is not supported by this provider");
            }
            return { exists: await provider.sqlite.exists() };
        },
    };
}

function toSessionFsError(err: unknown): SessionFsError {
    const e = err as NodeJS.ErrnoException;
    const code = e.code === "ENOENT" ? "ENOENT" : "UNKNOWN";
    return { code, message: e.message ?? String(err) };
}

function toSqliteTransactionError(err: unknown): GeneratedSqliteTransactionError {
    if (err instanceof SessionFsSqliteTransactionFailure) {
        return { errorClass: err.errorClass, message: err.message };
    }
    return {
        errorClass: "fatal",
        message: err instanceof Error ? err.message : String(err),
    };
}

File: ➡️📂Core kernel Yaiwes/GitHub-Copilot-SDK/GitHub-Copilot-SDK/dotnet/src/SessionFsProvider.cs
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

using GitHub.Copilot.Rpc;
using System.Diagnostics.CodeAnalysis;
using System.Text.Json;

namespace GitHub.Copilot;

/// <summary>
/// Result of a SQLite query execution via <see cref="ISessionFsSqliteProvider"/>.
/// Same shape as <see cref="SessionFsSqliteQueryResult"/> but without the <c>Error</c> field,
/// since providers signal errors by throwing.
/// </summary>
public sealed class SessionFsSqliteResult
{
    /// <summary>Column names from the result set.</summary>
    public IList<string> Columns { get; set; } = [];

    /// <summary>For SELECT: rows as column-keyed dictionaries. For others: empty.</summary>
    public IList<IDictionary<string, object>> Rows { get; set; } = [];

    /// <summary>Number of rows affected (for INSERT/UPDATE/DELETE).</summary>
    public long RowsAffected { get; set; }

    /// <summary>Last inserted row ID (for INSERT).</summary>
    public long? LastInsertRowid { get; set; }
}

/// <summary>
/// One statement in an atomic SQLite transaction passed to
/// <see cref="ISessionFsSqliteTransactionProvider.TransactionAsync"/>.
/// </summary>
[Experimental(Diagnostics.Experimental)]
public sealed class SessionFsSqliteStatement
{
    /// <summary>How to execute: <c>"exec"</c>, <c>"query"</c>, or <c>"run"</c>.</summary>
    public SessionFsSqliteQueryType QueryType { get; set; }

    /// <summary>SQL statement to execute.</summary>
    public string Query { get; set; } = string.Empty;

    /// <summary>Optional named bind parameters.</summary>
    public IDictionary<string, object?>? Params { get; set; }
}

/// <summary>
/// Optional interface for <see cref="SessionFsProvider"/> subclasses that support
/// per-session SQLite databases. Implement this interface on your provider to enable
/// the runtime's SQL tool to route queries through your SessionFs implementation.
/// </summary>
public interface ISessionFsSqliteProvider
{
    /// <summary>
    /// Executes a SQLite query against the per-session database.
    /// </summary>
    /// <param name="queryType">How to execute: <c>"exec"</c> for DDL/multi-statement, <c>"query"</c> for SELECT, <c>"run"</c> for INSERT/UPDATE/DELETE.</param>
    /// <param name="query">SQL query to execute.</param>
    /// <param name="bindParams">Optional named bind parameters.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The query result, or <c>null</c> for exec-type queries.</returns>
    Task<SessionFsSqliteResult?> QueryAsync(
        SessionFsSqliteQueryType queryType,
        string query,
        IDictionary<string, object?>? bindParams,
        CancellationToken cancellationToken);

    /// <summary>
    /// Checks whether the per-session SQLite database already exists, without creating it.
    /// </summary>
    /// <param name="cancellationToken">Cancellation token.</param>
    Task<bool> ExistsAsync(CancellationToken cancellationToken);
}

/// <summary>
/// Optional capability for session filesystem providers that support atomic SQLite transactions.
/// </summary>
public interface ISessionFsSqliteTransactionProvider
{
    /// <summary>
    /// Executes <paramref name="statements"/> atomically against the per-session database.
    /// </summary>
    /// <param name="statements">Statements to execute in order, inside a single transaction.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>One result per statement, in the same order as <paramref name="statements"/>.</returns>
    /// <exception cref="SessionFsSqliteTransactionException">
    /// Thrown to tell the runtime how the failure should be classified. Any other exception
    /// is reported as <see cref="SessionFsSqliteTransactionErrorClass.Fatal"/>.
    /// </exception>
    Task<IList<SessionFsSqliteResult>> TransactionAsync(
        IList<SessionFsSqliteStatement> statements,
        CancellationToken cancellationToken);
}

/// <summary>
/// Thrown by an <see cref="ISessionFsSqliteTransactionProvider"/> to classify a failed SQLite transaction.
/// <see cref="SessionFsSqliteTransactionErrorClass.BusyOrLocked"/> guarantees the transaction
/// rolled back and is safe to retry; <see cref="SessionFsSqliteTransactionErrorClass.PostCommitAmbiguous"/>
/// must never be retried.
/// </summary>
[Experimental(Diagnostics.Experimental)]
public sealed class SessionFsSqliteTransactionException : Exception
{
    /// <summary>Initializes a new instance of the <see cref="SessionFsSqliteTransactionException"/> class.</summary>
    /// <param name="message">Human-readable failure description.</param>
    /// <param name="errorClass">How the runtime should classify the failure.</param>
    /// <param name="innerException">Optional underlying exception.</param>
    public SessionFsSqliteTransactionException(
        string message,
        SessionFsSqliteTransactionErrorClass errorClass,
        Exception? innerException = null)
        : base(message, innerException)
    {
        ErrorClass = errorClass;
    }

    /// <summary>Gets the failure classification reported to the runtime.</summary>
    public SessionFsSqliteTransactionErrorClass ErrorClass { get; }
}

/// <summary>
/// Base class for session filesystem providers. Subclasses override the
/// virtual methods and use normal C# patterns (return values, throw exceptions).
/// The base class catches exceptions and converts them to <see cref="SessionFsError"/>
/// results expected by the runtime.
/// To add SQLite support, also implement <see cref="ISessionFsSqliteProvider"/>.
/// </summary>
public abstract class SessionFsProvider : ISessionFsHandler
{
    /// <summary>Reads the full content of a file. Throw if the file does not exist.</summary>
    /// <param name="path">SessionFs-relative path.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The file content as a UTF-8 string.</returns>
    protected abstract Task<string> ReadFileAsync(string path, CancellationToken cancellationToken);

    /// <summary>Writes content to a file, creating it (and parent directories) if needed.</summary>
    /// <param name="path">SessionFs-relative path.</param>
    /// <param name="content">Content to write.</param>
    /// <param name="mode">Optional POSIX-style permission mode. Null means use OS default.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    protected abstract Task WriteFileAsync(string path, string content, int? mode, CancellationToken cancellationToken);

    /// <summary>Appends content to a file, creating it (and parent directories) if needed.</summary>
    /// <param name="path">SessionFs-relative path.</param>
    /// <param name="content">Content to append.</param>
    /// <param name="mode">Optional POSIX-style permission mode. Null means use OS default.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    protected abstract Task AppendFileAsync(string path, string content, int? mode, CancellationToken cancellationToken);

    /// <summary>Checks whether a path exists.</summary>
    /// <param name="path">SessionFs-relative path.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns><c>true</c> if the path exists, <c>false</c> otherwise.</returns>
    protected abstract Task<bool> ExistsAsync(string path, CancellationToken cancellationToken);

    /// <summary>Gets metadata about a file or directory. Throw if the path does not exist.</summary>
    /// <param name="path">SessionFs-relative path.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    protected abstract Task<SessionFsStatResult> StatAsync(string path, CancellationToken cancellationToken);

    /// <summary>Creates a directory (and optionally parents). Does not fail if it already exists.</summary>
    /// <param name="path">SessionFs-relative path.</param>
    /// <param name="recursive">Whether to create parent directories.</param>
    /// <param name="mode">Optional POSIX-style permission mode (e.g., 0x1FF for 0777). Null means use OS default.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    protected abstract Task MakeDirectoryAsync(string path, bool recursive, int? mode, CancellationToken cancellationToken);

    /// <summary>Lists entry names in a directory. Throw if the directory does not exist.</summary>
    /// <param name="path">SessionFs-relative path.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    protected abstract Task<IList<string>> ReadDirectoryAsync(string path, CancellationToken cancellationToken);

    /// <summary>Lists entries with type info in a directory. Throw if the directory does not exist.</summary>
    /// <param name="path">SessionFs-relative path.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    protected abstract Task<IList<SessionFsReaddirWithTypesEntry>> ReadDirectoryWithTypesAsync(string path, CancellationToken cancellationToken);

    /// <summary>Removes a file or directory. Throw if the path does not exist (unless <paramref name="force"/> is true).</summary>
    /// <param name="path">SessionFs-relative path.</param>
    /// <param name="recursive">Whether to remove directory contents recursively.</param>
    /// <param name="force">If true, do not throw when the path does not exist.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    protected abstract Task RemoveAsync(string path, bool recursive, bool force, CancellationToken cancellationToken);

    /// <summary>Renames/moves a file or directory.</summary>
    /// <param name="src">Source path.</param>
    /// <param name="dest">Destination path.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    protected abstract Task RenameAsync(string src, string dest, CancellationToken cancellationToken);

    // ---- ISessionFsHandler implementation (private, handles error mapping) ----

    async Task<SessionFsReadFileResult> ISessionFsHandler.ReadFileAsync(SessionFsReadFileRequest request, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(request);

        try
        {
            var content = await ReadFileAsync(request.Path, cancellationToken).ConfigureAwait(false);
            return new SessionFsReadFileResult { Content = content };
        }
        catch (Exception ex)
        {
            return new SessionFsReadFileResult { Error = ToSessionFsError(ex) };
        }
    }

    async Task<SessionFsError?> ISessionFsHandler.WriteFileAsync(SessionFsWriteFileRequest request, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(request);

        try
        {
            await WriteFileAsync(request.Path, request.Content, (int?)request.Mode, cancellationToken).ConfigureAwait(false);
            return null;
        }
        catch (Exception ex)
        {
            return ToSessionFsError(ex);
        }
    }

    async Task<SessionFsError?> ISessionFsHandler.AppendFileAsync(SessionFsAppendFileRequest request, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(request);

        try
        {
            await AppendFileAsync(request.Path, request.Content, (int?)request.Mode, cancellationToken).ConfigureAwait(false);
            return null;
        }
        catch (Exception ex)
        {
            return ToSessionFsError(ex);
        }
    }

    async Task<SessionFsExistsResult> ISessionFsHandler.ExistsAsync(SessionFsExistsRequest request, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(request);

        try
        {
            var exists = await ExistsAsync(request.Path, cancellationToken).ConfigureAwait(false);
            return new SessionFsExistsResult { Exists = exists };
        }
        catch
        {
            return new SessionFsExistsResult { Exists = false };
        }
    }

    async Task<SessionFsStatResult> ISessionFsHandler.StatAsync(SessionFsStatRequest request, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(request);

        try
        {
            return await StatAsync(request.Path, cancellationToken).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            return new SessionFsStatResult { Error = ToSessionFsError(ex) };
        }
    }

    async Task<SessionFsError?> ISessionFsHandler.MkdirAsync(SessionFsMkdirRequest request, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(request);

        try
        {
            await MakeDirectoryAsync(request.Path, request.Recursive ?? false, (int?)request.Mode, cancellationToken).ConfigureAwait(false);
            return null;
        }
        catch (Exception ex)
        {
            return ToSessionFsError(ex);
        }
    }

    async Task<SessionFsReaddirResult> ISessionFsHandler.ReaddirAsync(SessionFsReaddirRequest request, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(request);

        try
        {
            var entries = await ReadDirectoryAsync(request.Path, cancellationToken).ConfigureAwait(false);
            return new SessionFsReaddirResult { Entries = entries };
        }
        catch (Exception ex)
        {
            return new SessionFsReaddirResult { Error = ToSessionFsError(ex) };
        }
    }

    async Task<SessionFsReaddirWithTypesResult> ISessionFsHandler.ReaddirWithTypesAsync(SessionFsReaddirWithTypesRequest request, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(request);

        try
        {
            var entries = await ReadDirectoryWithTypesAsync(request.Path, cancellationToken).ConfigureAwait(false);
            return new SessionFsReaddirWithTypesResult { Entries = entries };
        }
        catch (Exception ex)
        {
            return new SessionFsReaddirWithTypesResult { Error = ToSessionFsError(ex) };
        }
    }

    async Task<SessionFsError?> ISessionFsHandler.RmAsync(SessionFsRmRequest request, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(request);

        try
        {
            await RemoveAsync(request.Path, request.Recursive ?? false, request.Force ?? false, cancellationToken).ConfigureAwait(false);
            return null;
        }
        catch (Exception ex)
        {
            return ToSessionFsError(ex);
        }
    }

    async Task<SessionFsError?> ISessionFsHandler.RenameAsync(SessionFsRenameRequest request, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(request);

        try
        {
            await RenameAsync(request.Src, request.Dest, cancellationToken).ConfigureAwait(false);
            return null;
        }
        catch (Exception ex)
        {
            return ToSessionFsError(ex);
        }
    }

    async Task<SessionFsSqliteQueryResult> ISessionFsHandler.SqliteQueryAsync(SessionFsSqliteQueryRequest request, CancellationToken cancellationToken)
    {
        if (this is not ISessionFsSqliteProvider sqliteProvider)
        {
            return new SessionFsSqliteQueryResult
            {
                Error = new SessionFsError { Code = SessionFsErrorCode.UNKNOWN, Message = "SQLite is not supported by this provider." },
            };
        }

        try
        {
            var bindParams = request.Params?.ToDictionary(
                kvp => kvp.Key,
                kvp => JsonElementToValue(kvp.Value));
            var result = await sqliteProvider.QueryAsync(request.QueryType, request.Query, bindParams, cancellationToken).ConfigureAwait(false);

            return new SessionFsSqliteQueryResult
            {
                Rows = result?.Rows?.Select(row => (IDictionary<string, JsonElement>)row.ToDictionary(
                    kvp => kvp.Key,
                    kvp => ToJsonElement(kvp.Value))).ToList() ?? [],
                Columns = result?.Columns ?? [],
                RowsAffected = result?.RowsAffected ?? 0,
                LastInsertRowid = result?.LastInsertRowid,
            };
        }
        catch (Exception ex)
        {
            return new SessionFsSqliteQueryResult { Error = ToSessionFsError(ex) };
        }
    }

    async Task<SessionFsSqliteTransactionResult> ISessionFsHandler.SqliteTransactionAsync(SessionFsSqliteTransactionRequest request, CancellationToken cancellationToken)
    {
        if (this is not ISessionFsSqliteTransactionProvider transactionProvider)
        {
            return new SessionFsSqliteTransactionResult
            {
                Error = new SessionFsSqliteTransactionError
                {
                    ErrorClass = SessionFsSqliteTransactionErrorClass.Fatal,
                    Message = "SQLite is not supported by this provider.",
                },
            };
        }

        IList<SessionFsSqliteResult> results;
        try
        {
            var statements = request.Statements.Select(statement => new SessionFsSqliteStatement
            {
                QueryType = statement.QueryType,
                Query = statement.Query,
                Params = statement.Params?.ToDictionary(kvp => kvp.Key, kvp => JsonElementToValue(kvp.Value)),
            }).ToList();
            results = await transactionProvider.TransactionAsync(statements, cancellationToken).ConfigureAwait(false);
        }
        catch (SessionFsSqliteTransactionException ex)
        {
            return new SessionFsSqliteTransactionResult
            {
                Error = new SessionFsSqliteTransactionError { ErrorClass = ex.ErrorClass, Message = ex.Message },
            };
        }
        catch (Exception ex)
        {
            return new SessionFsSqliteTransactionResult
            {
                Error = new SessionFsSqliteTransactionError
                {
                    ErrorClass = SessionFsSqliteTransactionErrorClass.Fatal,
                    Message = ex.Message,
                },
            };
        }

        try
        {
            return new SessionFsSqliteTransactionResult
            {
                Results = results.Select(result => new SessionFsSqliteQueryResult
                {
                    Rows = result.Rows?.Select(row => (IDictionary<string, JsonElement>)row.ToDictionary(
                        kvp => kvp.Key,
                        kvp => ToJsonElement(kvp.Value))).ToList() ?? [],
                    Columns = result.Columns ?? [],
                    RowsAffected = result.RowsAffected,
                    LastInsertRowid = result.LastInsertRowid,
                }).ToList(),
            };
        }
        catch (Exception ex)
        {
            return new SessionFsSqliteTransactionResult
            {
                Error = new SessionFsSqliteTransactionError
                {
                    ErrorClass = SessionFsSqliteTransactionErrorClass.PostCommitAmbiguous,
                    Message = ex.Message,
                },
            };
        }
    }

    async Task<SessionFsSqliteExistsResult> ISessionFsHandler.SqliteExistsAsync(SessionFsSqliteExistsRequest request, CancellationToken cancellationToken)
    {
        if (this is not ISessionFsSqliteProvider sqliteProvider)
        {
            return new SessionFsSqliteExistsResult { Exists = false };
        }

        try
        {
            var exists = await sqliteProvider.ExistsAsync(cancellationToken).ConfigureAwait(false);
            return new SessionFsSqliteExistsResult { Exists = exists };
        }
        catch
        {
            return new SessionFsSqliteExistsResult { Exists = false };
        }
    }


    private static SessionFsError ToSessionFsError(Exception ex)
    {
        var code = ex is FileNotFoundException or DirectoryNotFoundException
            ? SessionFsErrorCode.ENOENT
            : SessionFsErrorCode.UNKNOWN;
        return new SessionFsError { Code = code, Message = ex.Message };
    }

    private static JsonElement ToJsonElement(object? value) =>
        CopilotClient.ToJsonElementForWire(value) ?? JsonElement.Parse("null");

    private static object? JsonElementToValue(JsonElement element) => element.ValueKind switch
    {
        JsonValueKind.Null => null,
        JsonValueKind.True => true,
        JsonValueKind.False => false,
        JsonValueKind.String => element.GetString(),
        JsonValueKind.Number => element.TryGetInt64(out var l) ? l : element.GetDouble(),
        _ => element.GetRawText(),
    };
}
